pkg/abcupdater/metrics.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package abcupdater
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"regexp"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/sethvargo/go-envconfig"
+
+	"github.com/abcxyz/abc-updater/pkg/abcupdater/localstore"
+	"github.com/abcxyz/pkg/logging"
+)
+
+const (
+	installIDFileName = "id.json"
+)
+
+var regExUUID = regexp.MustCompile("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")
+
+type metricsConfig struct {
+	ServerURL string `env:"ABC_UPDATER_METRICS_URL,default=https://abc-updater-metrics.tycho.joonix.net"`
+}
+
+// MetricsInfo are the parameters for sending metrics.
+type MetricsInfo struct {
+	// The ID of the application to check.
+	AppID string
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	Version string
+
+	// An optional Lookuper to load envconfig structs. Will default to os environment variables.
+	Lookuper envconfig.Lookuper
+
+	Metrics map[string]int
+
+	// Optional override for install id file location. Mostly intended for testing.
+	// If empty uses default location.
+	InstallIDFileOverride string
+}
+
+// InstallIDData defines the json file that defines installation id.
+type InstallIDData struct {
+	// Time ID was created, in UTC epoch seconds.
+	IDCreatedTimestamp int64 `json:"idCreatedTimestamp"`
+	// InstallID. Expected to be a hex 8-4-4-4-12 formatted v4 UUID.
+	InstallID string `json:"installId"`
+}
+
+type SendMetricRequest struct {
+	// The ID of the application to check.
+	AppID string `json:"appId"`
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	Version string `json:"version"`
+
+	Metrics map[string]int `json:"metrics"`
+
+	// InstallID. Expected to be a hex 8-4-4-4-12 formatted v4 UUID.
+	InstallID string `json:"installId"`
+}
+
+// Stricter than uuid.Parse() which isn't meant for validating strings according
+// to documentation.
+func validInstallID(id string) bool {
+	return regExUUID.MatchString(id)
+}
+
+// SendMetricsSync sends information about application usage. Users can opt out
+// by setting an env variable defined in opt_out.go.
+// Accepts a context for cancellation.
+func SendMetricsSync(ctx context.Context, info *MetricsInfo) error {
```

Comment:
Rather than doing N loads of the config file to send N events, I wonder if we want to load the config once and reuse it. Maybe you'd have a `type Reporter struct` that has a `Send` method, or something?

Reply:
I guess my thought was doing one call per application run, as metrics are meant to be pretty basic (`abc` was ran with the `render` sub command).

We could talk about making it more interactive though